### ISSUE TITLE
unresolved-attribute for variables that start as "self.xxx = None"

### ISSUE THREAD

**Issue by npaolini2634** [astral-sh/ty#2839]
Starting in ty version 0.0.17 errors started popping up for instance attributes that start out as "None" and are assigned later.

```
class Test:
    def __init__(self):
        self.xxx = None
        self.fn_1()
        self.fn_2()

    def fn_1(self) -> None:
        self.xxx = 'populated'

    def fn_2(self) -> None:
        print(self.xxx.strip())  # <- Error if "unresolved-attribute" is enabled


# If "unresolved-attribute" is disabled, I won't see this real error
yyy = 1
print(yyy.strip())
```

I know it's possible to avoid these errors by performing a `if self.xxx:` check before using `self.xxx`, but for some scripts that involves littering the code with `if self.xxx:` checks, for little added value. It would be nice to have a rule to control this. Pyright has configuration flags such as "reportOptionalMemberAccess" to allow for this behavior, and mypy has "strict_optional".

Still loving using ty, and appreciate all your hard work!

**Comment by carljm** [comment]
Thanks for the report!

We could add an equivalent to `reportOptionalMemberAccess`. `strict_optional` is even more lax, I think, as it suppresses all union-with-None errors, not just ones from accessing attributes.

Personally, the workaround I would probably prefer over using either of those options would be to add `xxx: str` annotation in the body of `class Test:` (to inform the type checker that it should treat the attribute as having type `str`), and then add a `# ty: ignore[invalid-assignment] always initialized to non-None later` comment on the `self.xxx = None` line.

If your real example does involve a case where it's always initialized to a non-None value later, what purpose does the `None` initialization serve? If it were eliminated, ty wouldn't infer an optional type for the attribute in the first place.

**Comment by carljm** [comment]
The thing that changed in 0.0.17 was that we intentionally made accessing an attribute on a union that doesn't exist on some members of the union into an `unresolved-attribute` error instead of a `possibly-missing-attribute` warning. `possibly-missing-attribute` is intended for cases where an attribute is conditionally defined on a class.

**Comment by npaolini2634** [comment]
Okay, good to know about the update logic for 0.0.17.

Side note (this may be a separate error), I think the invalid-assignment error should come up like you mentioned in your suggestion, but it doesn't seem to.

```py
abc: int = 'no'  # invalid-assignment

class Test:
    def __init__(self):
        self.xxx: str = None  # no error. Should be invalid-assignment?
```

**To your questions "what purpose does the None initialization serve?"**
I've always assumed it was best practice to initialize all self variables in the `__init__` of a class. This provides a quick reference for what variables a class has. I think PyCharm even has a warning by default if you don't do this. 

**Accessing attributes on optional variables**
I agree that in some cases, especially the black-and-white one like I described, doing `self.xxx: str = None  # ty: ignore` could work. There are lots of other more gray situations where it would be more awkward or not possible to do that, and the fix then becomes adding more conditions or asserts to the code to satisfy the type checker.

I'm working with a large repo that used to have no type checking. I did a huge amount of work to get type checking working on it, but currently there would be 930 more errors I'd need to fix if there wasn't a flag to allow accessing attributes on optional variables. If a flag isn't added for this, I might just go through and update everything, but I think there are a lot of programmers that in my position would instead just switch back to using mypy or Pyright. So overall, I think adoption of type checking on non type checked repos will be much harder if a flag like this doesn't exist.

**Comment by carljm** [comment]
`self.xxx: str = None` should be an invalid-assignment error, yeah - that's a bug. Thanks for reporting it! Filed https://github.com/astral-sh/ty/issues/2840 to track that.

> I've always assumed it was best practice to initialize all self variables in the `__init__` of a class. This provides a quick reference for what variables a class has. I think PyCharm even has a warning by default if you don't do this.

Providing class-level annotations `xxx: str` can provide an even easier-to-scan reference. But yeah, I see the rationale. Thanks!

**Comment by AlexWaygood** [comment]
I dislike the idea of treating an `X | None` union any differently to an `X | Y` union where `Y` is any other type. Unions with `None` are not treated differently in the type system, and attribute access on `None` is not treated any differently at runtime.

However, a compromise here might be to have a dedicated error code (neither `possibly-missing-attribute` nor `unresolved-attribute`, but a third thing!) for invalid access on union types (including unions with `None`), similar to mypy's [union-attr](https://mypy.readthedocs.io/en/stable/error_code_list.html#check-that-attribute-exists-in-each-union-item-union-attr) error code. Then:
- We could have it enabled with error severity by default, as it should be; but
- If users find it very noisy they would be able to disable it for certain files, or even globally.

**Comment by Otto-AA** [comment]
FYI, PyCharms inbuilt type checking also treats unions as "if any of the union types matches, it's okay" rather than the normal "if all of the union types match, it's okay" (e.g. if `x: str | int` it will happily allow `x.upper()`). That's not the best for type safety, but I guess such an option could be useful when migrating from the PyCharm type checker.

**Comment by carljm** [comment]
Mypy seems to also emit `union-attr` when _no_ element of the union has the attribute. IMO this makes it less useful as a separate error code. I think we'd want to only use the dedicated error code when some element(s) do have it and others do not (i.e. the cases that would pass in PyCharm).

**Comment by AlexWaygood** [comment]
> Mypy seems to also emit `union-attr` when _no_ element of the union has the attribute. IMO this makes it less useful as a separate error code. I think we'd want to only use the dedicated error code when some element(s) do have it and others do not (i.e. the cases that would pass in PyCharm).

agreed

**Comment by gryznar** [comment]
@AlexWaygood issue which I've previously submitted had much broader scope than just attribute access. Nullability may also apply for e.g. `invalid-argument-type` and many other rules. I'd appreciate solution which will be more comprehensive. 

As I suppose, this cannot be solved by just adding rule for any union

**Comment by AlexWaygood** [comment]
In general, if something could be `None` and you haven't accounted for that in your code, that's a serious typing error and something I'd personally prioritise fixing. I'm sympathetic to adding a different error code for attribute access specifically, especially since it follows the pattern of other type checkers. But I don't think we'll be making a broader change here, I'm afraid.

**Comment by ItsCubeTime** [comment]
> In general, if something could be None and you haven't accounted for that in your code, that's a serious typing error

@AlexWaygood Consider the following logic:
```py
class Foo:
    def __init__(self,hasBar:bool): self.hasBar=hasBar;i hasBar:self.bar="qux" if hasBar else None
print(Foo(True).bar)
```
The API I work against `bpy`, Blender's ( https://www.blender.org ) uses similar syntax extensively to help determine types.

When using the API, you are expected to use logic like:
```py
foo=Foo(True) # Well, objects are not really constructed like this, but serves the purposes of examplification here
if foo.hasBar:
    doSomething(foo.bar)
```

In many cases, its the underlying C++ codebase of the project that controls the state that drives `hasBar`, so there would be no way for a static analyzer like `ty` to tell if `bar` exists, or is of a particular type under the particular conditions that the code is executed, something that only a person understanding those conditions will understand (runtime dependent).

To provide a straight forward example using `bpy`:
```py
import bpy
for obj in bpy.context.selected_objects:
    if obj.type == 'MESH':
        mesh:bpy.types.Mesh=obj.data
```
<img width="1651" height="78" alt="Image" src="https://github.com/user-attachments/assets/b1fa8df7-e7aa-43b2-af84-8a551fa254bc" />

With that, I would like an option to make Ty just assume that the first type it finds that has the attribute is the one you intend to use (for with the way the API is designed, I see no other way); PyLance does something in this direction I believe?

**Comment by ItsCubeTime** [comment]
On a second thought, in the case of `bpy`, perhaps disabling the error is not actually what one may need/want to do, but rather have a setting to make Ty look for attribtes present on any of the types present in the `Union` (instead of attributes present on all)

**Comment by carljm** [comment]
@ItsCubeTime I think addressing this at the level of attribute handling is the wrong level. You may also want to pass the object to a function that requires a specific subtype with some attribute, for example. The core problem here is that you want `if foo.hasBar` to narrow the type of `foo` from a broad union to a more precise type, but this isn't supported in the Python type system. You could write a function `has_bar(x: Foo) -> TypeIs[HasBar]` and then write `if has_bar(foo)` instead of `if foo.has_bar`, and this could give you the narrowing you want.
